File: Tests/DemoTapeTests/DemoControlTests.swift
import XCTest
@testable import DemoTape

final class DemoControlTests: XCTestCase {

    private func parse(_ s: String) -> DemoControl.Command? {
        guard let url = URL(string: s) else { return nil }
        return DemoControl.parse(url)
    }

    func testStopVariants() {
        XCTAssertEqual(parse("demotape://record/stop"), .stop)
        XCTAssertEqual(parse("demotape://stop"), .stop)
        XCTAssertEqual(parse("DEMOTAPE://record/STOP"), .stop)
    }

    func testStartFullScreenDefaults() {
        guard case .start(let opts)? = parse("demotape://record/start") else { return XCTFail() }
        XCTAssertEqual(opts.region, .fullScreen)
        XCTAssertEqual(opts.countdown, 3)
        XCTAssertNil(opts.microphone)
        XCTAssertNil(opts.webcam)
    }

    func testStartImmediate() {
        guard case .start(let opts)? = parse("demotape://record/start?countdown=0") else { return XCTFail() }
        XCTAssertEqual(opts.countdown, 0)
    }

    func testStartNormalizedRegion() {
        guard case .start(let opts)? = parse("demotape://record/start?nx=0.1&ny=0.2&nw=0.5&nh=0.4") else { return XCTFail() }
        XCTAssertEqual(opts.region, .normalized(CGRect(x: 0.1, y: 0.2, width: 0.5, height: 0.4)))
    }

    func testStartPixelRegion() {
        guard case .start(let opts)? = parse("demotape://record/start?mode=area&x=100&y=80&w=1280&h=720") else { return XCTFail() }
        XCTAssertEqual(opts.region, .pixels(CGRect(x: 100, y: 80, width: 1280, height: 720)))
    }

    func testStartInputFlags() {
        guard case .start(let opts)? = parse("demotape://record/start?mic=1&webcam=0") else { return XCTFail() }
        XCTAssertEqual(opts.microphone, true)
        XCTAssertEqual(opts.webcam, false)
    }

    func testNormalizedTakesPrecedenceOverPixels() {
        guard case .start(let opts)? = parse("demotape://record/start?nx=0&ny=0&nw=1&nh=1&x=5&y=5&w=5&h=5") else { return XCTFail() }
        XCTAssertEqual(opts.region, .normalized(CGRect(x: 0, y: 0, width: 1, height: 1)))
    }

    func testRejectsForeignSchemeAndGarbage() {
        XCTAssertNil(parse("https://record/start"))
        XCTAssertNil(parse("demotape://record/pause"))
        XCTAssertNil(parse("demotape://"))
    }

    func testCursorMove() {
        XCTAssertEqual(parse("demotape://cursor/move?x=640&y=360"),
                       .cursor(x: 640, y: 360, click: false))
    }

    func testCursorClick() {
        XCTAssertEqual(parse("demotape://cursor/click?x=12&y=34"),
                       .cursor(x: 12, y: 34, click: true))
    }

    func testCursorRequiresCoordinates() {
        XCTAssertNil(parse("demotape://cursor/click"))
        XCTAssertNil(parse("demotape://cursor/move?x=10"))
    }
}

/// `demotape://ui/…` — lets a scripted walkthrough open DemoTape's own windows, since a menu row's
/// screen rect isn't discoverable from outside the app.
extension DemoControlTests {

    func testParseOpenUIByQuery() {
        let url = URL(string: "demotape://ui/open?window=publish")!
        XCTAssertEqual(DemoControl.parse(url), .openUI(.publish))
    }

    func testParseOpenUIShorthandPath() {
        let url = URL(string: "demotape://ui/about")!
        XCTAssertEqual(DemoControl.parse(url), .openUI(.about))
    }

    func testParseOpenUIIsCaseInsensitive() {
        let url = URL(string: "demotape://ui/open?window=COMPOSER")!
        XCTAssertEqual(DemoControl.parse(url), .openUI(.composer))
    }

    func testParseOpenUIRejectsUnknownWindow() {
        let url = URL(string: "demotape://ui/open?window=nope")!
        XCTAssertNil(DemoControl.parse(url))
    }

    func testEveryWindowCaseParses() {
        for w in DemoControl.Window.allCases {
            let url = URL(string: "demotape://ui/open?window=\(w.rawValue)")!
            XCTAssertEqual(DemoControl.parse(url), .openUI(w), "failed for \(w.rawValue)")
        }
    }
}

/// Cursor glide — animated, human-looking travel instead of a teleport.
extension DemoControlTests {

    func testPlainCursorDoesNotGlide() {
        let url = URL(string: "demotape://cursor?x=100&y=200")!
        XCTAssertEqual(DemoControl.parse(url), .cursor(x: 100, y: 200, click: false, glideMs: 0))
    }

    func testGlidePathGetsDefaultDuration() {
        let url = URL(string: "demotape://cursor/glide?x=100&y=200")!
        XCTAssertEqual(DemoControl.parse(url), .cursor(x: 100, y: 200, click: false, glideMs: 420))
    }

    func testExplicitMsOverridesGlideDefault() {
        let url = URL(string: "demotape://cursor/glide?x=10&y=20&ms=900")!
        XCTAssertEqual(DemoControl.parse(url), .cursor(x: 10, y: 20, click: false, glideMs: 900))
    }

    func testMsWorksWithoutGlidePathSegment() {
        let url = URL(string: "demotape://cursor?x=10&y=20&ms=250")!
        XCTAssertEqual(DemoControl.parse(url), .cursor(x: 10, y: 20, click: false, glideMs: 250))
    }

    func testGlideAndClickCombine() {
        let url = URL(string: "demotape://cursor/glide/click?x=5&y=6&ms=300")!
        XCTAssertEqual(DemoControl.parse(url), .cursor(x: 5, y: 6, click: true, glideMs: 300))
    }

    func testNegativeDurationIsClampedToZero() {
        let url = URL(string: "demotape://cursor?x=1&y=2&ms=-500")!
        XCTAssertEqual(DemoControl.parse(url), .cursor(x: 1, y: 2, click: false, glideMs: 0))
    }

    func testCursorStillRequiresBothCoordinates() {
        XCTAssertNil(DemoControl.parse(URL(string: "demotape://cursor/glide?x=10")!))
    }
}

/// Semantic (coordinate-free) targeting through the control surface.
extension DemoControlTests {

    func testParseElementClickByLabel() {
        let url = URL(string: "demotape://ui/click?label=Export")!
        XCTAssertEqual(DemoControl.parse(url),
                       .element(query: .init(label: "Export"), click: true))
    }

    func testParseElementFindDoesNotClick() {
        let url = URL(string: "demotape://ui/find?label=Export")!
        XCTAssertEqual(DemoControl.parse(url),
                       .element(query: .init(label: "Export"), click: false))
    }

    func testParseElementCarriesRoleAppAndIndex() {
        let url = URL(string: "demotape://ui/click?label=Allow&role=AXButton&app=Safari&index=2")!
        XCTAssertEqual(DemoControl.parse(url),
                       .element(query: .init(label: "Allow", role: "AXButton",
                                             app: "Safari", index: 2), click: true))
    }

    func testParseElementRequiresALabel() {
        XCTAssertNil(DemoControl.parse(URL(string: "demotape://ui/click?role=AXButton")!))
    }

    func testParseDumpUI() {
        XCTAssertEqual(DemoControl.parse(URL(string: "demotape://ui/dump")!), .dumpUI(app: nil))
        XCTAssertEqual(DemoControl.parse(URL(string: "demotape://ui/dump?app=Safari")!),
                       .dumpUI(app: "Safari"))
    }
}

/// `+` in a query value means a space — scripts build these URLs by hand.
extension DemoControlTests {

    func testPlusIsDecodedAsSpaceInLabels() {
        let url = URL(string: "demotape://ui/click?label=Check+for+Updates")!
        XCTAssertEqual(DemoControl.parse(url),
                       .element(query: .init(label: "Check for Updates"), click: true))
    }

    func testPercentEncodedSpacesStillWork() {
        let url = URL(string: "demotape://ui/click?label=Check%20for%20Updates")!
        XCTAssertEqual(DemoControl.parse(url),
                       .element(query: .init(label: "Check for Updates"), click: true))
    }

    func testPlusDecodingAppliesToAppNames() {
        let url = URL(string: "demotape://ui/dump?app=Visual+Studio+Code")!
        XCTAssertEqual(DemoControl.parse(url), .dumpUI(app: "Visual Studio Code"))
    }
}

/// The menu must be told how long to stay open: while it tracks, it blocks every later command.
extension DemoControlTests {

    func testMenuHoldIsParsed() {
        let url = URL(string: "demotape://ui/open?window=menu&hold=2500")!
        XCTAssertEqual(DemoControl.parse(url), .openUI(.menu, holdMs: 2500))
    }

    func testHoldDefaultsToZeroMeaningNoAutoDismiss() {
        let url = URL(string: "demotape://ui/open?window=menu")!
        XCTAssertEqual(DemoControl.parse(url), .openUI(.menu, holdMs: 0))
    }

    func testNegativeHoldIsClamped() {
        let url = URL(string: "demotape://ui/open?window=menu&hold=-400")!
        XCTAssertEqual(DemoControl.parse(url), .openUI(.menu, holdMs: 0))
    }
}

/// Real OS typing. Auto-zoom is driven by clicks AND keys, so typing the app can't observe leaves
/// the camera wide while text appears.
extension DemoControlTests {

    func testParseTypeText() {
        let url = URL(string: "demotape://type?text=hello%20world")!
        XCTAssertEqual(DemoControl.parse(url), .type(text: "hello world", cps: 0, expectedApp: nil))
    }

    func testParseTypeWithRate() {
        let url = URL(string: "demotape://type?text=abc&cps=12.5")!
        XCTAssertEqual(DemoControl.parse(url), .type(text: "abc", cps: 12.5, expectedApp: nil))
    }

    func testParseTypeDecodesPlusAsSpace() {
        let url = URL(string: "demotape://type?text=roll+it+back")!
        XCTAssertEqual(DemoControl.parse(url), .type(text: "roll it back", cps: 0, expectedApp: nil))
    }

    func testParseTypeRequiresText() {
        XCTAssertNil(DemoControl.parse(URL(string: "demotape://type?cps=10")!))
        XCTAssertNil(DemoControl.parse(URL(string: "demotape://type?text=")!))
    }

    func testParseTypeClampsNegativeRate() {
        let url = URL(string: "demotape://type?text=abc&cps=-5")!
        XCTAssertEqual(DemoControl.parse(url), .type(text: "abc", cps: 0, expectedApp: nil))
    }
}

/// Typing ACTIVITY — records the zoom hold without posting keystrokes, for text a browser
/// automation tool types itself (browsers drop synthetic keys that carry no virtual keycode).
extension DemoControlTests {

    func testParseTypingActivity() {
        let url = URL(string: "demotape://typing?chars=42&cps=14")!
        XCTAssertEqual(DemoControl.parse(url), .typingActivity(chars: 42, cps: 14))
    }

    func testTypingActivityRateIsOptional() {
        let url = URL(string: "demotape://typing?chars=8")!
        XCTAssertEqual(DemoControl.parse(url), .typingActivity(chars: 8, cps: 0))
    }

    func testTypingActivityRequiresPositiveCount() {
        XCTAssertNil(DemoControl.parse(URL(string: "demotape://typing?chars=0")!))
        XCTAssertNil(DemoControl.parse(URL(string: "demotape://typing?cps=14")!))
    }

    /// `typing` (activity) and `type` (real keystrokes) must stay distinct commands.
    func testTypingActivityIsNotConfusedWithRealTyping() {
        XCTAssertEqual(DemoControl.parse(URL(string: "demotape://type?text=hi")!),
                       .type(text: "hi", cps: 0, expectedApp: nil))
        XCTAssertEqual(DemoControl.parse(URL(string: "demotape://typing?chars=2")!),
                       .typingActivity(chars: 2, cps: 0))
    }
}

/// The `app=` guard on real OS typing. Keystrokes follow SYSTEM focus, so a caller must be able to
/// say which app it expects to be frontmost — otherwise a mis-timed command types into the user's
/// editor, destructively and silently.
extension DemoControlTests {

    func testTypeCarriesExpectedApp() {
        let url = URL(string: "demotape://type?text=hi&app=Chromium")!
        XCTAssertEqual(DemoControl.parse(url), .type(text: "hi", cps: 0, expectedApp: "Chromium"))
    }

    func testExpectedAppSurvivesPlusDecoding() {
        let url = URL(string: "demotape://type?text=hi&app=Google+Chrome")!
        XCTAssertEqual(DemoControl.parse(url),
                       .type(text: "hi", cps: 0, expectedApp: "Google Chrome"))
    }

    func testExpectedAppIsOptional() {
        let url = URL(string: "demotape://type?text=hi")!
        XCTAssertEqual(DemoControl.parse(url), .type(text: "hi", cps: 0, expectedApp: nil))
    }
}

File: windows/tests/DemoTape.Tests/DemoControlTests.cs
using DemoTape.Domain.Control;
using Xunit;

namespace DemoTape.Tests;

/// <summary>
/// Ported from the macOS <c>DemoControlTests.swift</c>: exhaustive parsing of the <c>demotape://</c>
/// control surface. Pure logic, no GUI/network, so it runs with only the .NET SDK.
/// </summary>
public class DemoControlTests
{
    private static DemoControl.Command? Parse(string s) =>
        Uri.TryCreate(s, UriKind.Absolute, out var url) ? DemoControl.Parse(url) : null;

    [Fact]
    public void StopVariants()
    {
        Assert.Equal(DemoControl.CommandKind.Stop, Parse("demotape://record/stop")!.Kind);
        Assert.Equal(DemoControl.CommandKind.Stop, Parse("demotape://stop")!.Kind);
        Assert.Equal(DemoControl.CommandKind.Stop, Parse("DEMOTAPE://record/STOP")!.Kind);
    }

    [Fact]
    public void StartFullScreenDefaults()
    {
        var c = Parse("demotape://record/start");
        Assert.Equal(DemoControl.CommandKind.Start, c!.Kind);
        Assert.Equal(DemoControl.RegionKind.FullScreen, c.Start!.Region.Kind);
        Assert.Equal(3, c.Start.Countdown);
        Assert.Null(c.Start.Microphone);
        Assert.Null(c.Start.Webcam);
    }

    [Fact]
    public void StartImmediate()
    {
        var c = Parse("demotape://record/start?countdown=0");
        Assert.Equal(0, c!.Start!.Countdown);
    }

    [Fact]
    public void StartNormalizedRegion()
    {
        var c = Parse("demotape://record/start?nx=0.1&ny=0.2&nw=0.5&nh=0.4");
        Assert.Equal(DemoControl.Region.Normalized(0.1, 0.2, 0.5, 0.4), c!.Start!.Region);
    }

    [Fact]
    public void StartPixelRegion()
    {
        var c = Parse("demotape://record/start?mode=area&x=100&y=80&w=1280&h=720");
        Assert.Equal(DemoControl.Region.Pixels(100, 80, 1280, 720), c!.Start!.Region);
    }

    [Fact]
    public void StartInputFlags()
    {
        var c = Parse("demotape://record/start?mic=1&webcam=0");
        Assert.True(c!.Start!.Microphone);
        Assert.False(c.Start.Webcam);
    }

    [Fact]
    public void NormalizedTakesPrecedenceOverPixels()
    {
        var c = Parse("demotape://record/start?nx=0&ny=0&nw=1&nh=1&x=5&y=5&w=5&h=5");
        Assert.Equal(DemoControl.Region.Normalized(0, 0, 1, 1), c!.Start!.Region);
    }

    [Fact]
    public void WebcamOnlyRoute()
    {
        var c = Parse("demotape://record/webcam?countdown=3");
        Assert.True(c!.Start!.WebcamOnly);
        Assert.Equal(3, c.Start.Countdown);
    }

    [Fact]
    public void ScreenStartIsNotWebcamOnly()
    {
        // A webcam=1 QUERY on a screen recording turns on the PiP bubble, not webcam-only mode.
        var c = Parse("demotape://record/start?webcam=1");
        Assert.False(c!.Start!.WebcamOnly);
        Assert.True(c.Start.Webcam);
    }

    [Fact]
    public void RejectsForeignSchemeAndGarbage()
    {
        Assert.Null(Parse("https://record/start"));
        Assert.Null(Parse("demotape://record/pause"));
        Assert.Null(Parse("demotape://"));
    }

    [Fact]
    public void CursorMove()
    {
        var c = Parse("demotape://cursor/move?x=640&y=360");
        Assert.Equal(DemoControl.CommandKind.Cursor, c!.Kind);
        Assert.Equal(640, c.CursorX);
        Assert.Equal(360, c.CursorY);
        Assert.False(c.Click);
    }

    [Fact]
    public void CursorClick()
    {
        var c = Parse("demotape://cursor/click?x=12&y=34");
        Assert.Equal(12, c!.CursorX);
        Assert.Equal(34, c.CursorY);
        Assert.True(c.Click);
    }

    [Fact]
    public void CursorRequiresCoordinates()
    {
        Assert.Null(Parse("demotape://cursor/click"));
        Assert.Null(Parse("demotape://cursor/move?x=10"));
    }

    // --- ui/open ---

    [Fact]
    public void ParseOpenUiByQuery()
    {
        var c = Parse("demotape://ui/open?window=publish");
        Assert.Equal(DemoControl.CommandKind.OpenUi, c!.Kind);
        Assert.Equal(DemoControl.UiWindow.Publish, c.Window);
    }

    [Fact]
    public void ParseOpenUiShorthandPath()
    {
        var c = Parse("demotape://ui/about");
        Assert.Equal(DemoControl.UiWindow.About, c!.Window);
    }

    [Fact]
    public void ParseOpenUiIsCaseInsensitive()
    {
        var c = Parse("demotape://ui/open?window=COMPOSER");
        Assert.Equal(DemoControl.UiWindow.Composer, c!.Window);
    }

    [Fact]
    public void ParseOpenUiRejectsUnknownWindow()
    {
        Assert.Null(Parse("demotape://ui/open?window=nope"));
    }

    [Fact]
    public void EveryWindowCaseParses()
    {
        foreach (DemoControl.UiWindow w in Enum.GetValues<DemoControl.UiWindow>())
        {
            var c = Parse($"demotape://ui/open?window={w.ToString().ToLowerInvariant()}");
            Assert.Equal(w, c!.Window);
        }
    }

    // --- cursor glide ---

    [Fact]
    public void PlainCursorDoesNotGlide()
    {
        var c = Parse("demotape://cursor?x=100&y=200");
        Assert.Equal(0, c!.GlideMs);
    }

    [Fact]
    public void GlidePathGetsDefaultDuration()
    {
        var c = Parse("demotape://cursor/glide?x=100&y=200");
        Assert.Equal(420, c!.GlideMs);
    }

    [Fact]
    public void ExplicitMsOverridesGlideDefault()
    {
        var c = Parse("demotape://cursor/glide?x=10&y=20&ms=900");
        Assert.Equal(900, c!.GlideMs);
    }

    [Fact]
    public void MsWorksWithoutGlidePathSegment()
    {
        var c = Parse("demotape://cursor?x=10&y=20&ms=250");
        Assert.Equal(250, c!.GlideMs);
    }

    [Fact]
    public void GlideAndClickCombine()
    {
        var c = Parse("demotape://cursor/glide/click?x=5&y=6&ms=300");
        Assert.True(c!.Click);
        Assert.Equal(300, c.GlideMs);
    }

    [Fact]
    public void NegativeDurationIsClampedToZero()
    {
        var c = Parse("demotape://cursor?x=1&y=2&ms=-500");
        Assert.Equal(0, c!.GlideMs);
    }

    [Fact]
    public void CursorStillRequiresBothCoordinates()
    {
        Assert.Null(Parse("demotape://cursor/glide?x=10"));
    }

    // --- semantic targeting ---

    [Fact]
    public void ParseElementClickByLabel()
    {
        var c = Parse("demotape://ui/click?label=Export");
        Assert.Equal(DemoControl.CommandKind.Element, c!.Kind);
        Assert.Equal("Export", c.Query!.Label);
        Assert.True(c.Click);
    }

    [Fact]
    public void ParseElementFindDoesNotClick()
    {
        var c = Parse("demotape://ui/find?label=Export");
        Assert.Equal("Export", c!.Query!.Label);
        Assert.False(c.Click);
    }

    [Fact]
    public void ParseElementCarriesRoleAppAndIndex()
    {
        var c = Parse("demotape://ui/click?label=Allow&role=AXButton&app=Safari&index=2");
        Assert.Equal(new DemoControl.ElementQuery("Allow", "AXButton", "Safari", 2), c!.Query);
        Assert.True(c.Click);
    }

    [Fact]
    public void ParseElementRequiresALabel()
    {
        Assert.Null(Parse("demotape://ui/click?role=AXButton"));
    }

    [Fact]
    public void ParseDumpUi()
    {
        Assert.Equal(DemoControl.CommandKind.DumpUi, Parse("demotape://ui/dump")!.Kind);
        Assert.Null(Parse("demotape://ui/dump")!.App);
        Assert.Equal("Safari", Parse("demotape://ui/dump?app=Safari")!.App);
    }

    // --- + decoding ---

    [Fact]
    public void PlusIsDecodedAsSpaceInLabels()
    {
        var c = Parse("demotape://ui/click?label=Check+for+Updates");
        Assert.Equal("Check for Updates", c!.Query!.Label);
    }

    [Fact]
    public void PercentEncodedSpacesStillWork()
    {
        var c = Parse("demotape://ui/click?label=Check%20for%20Updates");
        Assert.Equal("Check for Updates", c!.Query!.Label);
    }

    [Fact]
    public void PlusDecodingAppliesToAppNames()
    {
        var c = Parse("demotape://ui/dump?app=Visual+Studio+Code");
        Assert.Equal("Visual Studio Code", c!.App);
    }

    // --- menu hold ---

    [Fact]
    public void MenuHoldIsParsed()
    {
        var c = Parse("demotape://ui/open?window=menu&hold=2500");
        Assert.Equal(DemoControl.UiWindow.Menu, c!.Window);
        Assert.Equal(2500, c.HoldMs);
    }

    [Fact]
    public void HoldDefaultsToZeroMeaningNoAutoDismiss()
    {
        Assert.Equal(0, Parse("demotape://ui/open?window=menu")!.HoldMs);
    }

    [Fact]
    public void NegativeHoldIsClamped()
    {
        Assert.Equal(0, Parse("demotape://ui/open?window=menu&hold=-400")!.HoldMs);
    }

    // --- real OS typing ---

    [Fact]
    public void ParseTypeText()
    {
        var c = Parse("demotape://type?text=hello%20world");
        Assert.Equal(DemoControl.CommandKind.Type, c!.Kind);
        Assert.Equal("hello world", c.Text);
        Assert.Equal(0, c.Cps);
        Assert.Null(c.ExpectedApp);
    }

    [Fact]
    public void ParseTypeWithRate()
    {
        var c = Parse("demotape://type?text=abc&cps=12.5");
        Assert.Equal("abc", c!.Text);
        Assert.Equal(12.5, c.Cps);
    }

    [Fact]
    public void ParseTypeDecodesPlusAsSpace()
    {
        var c = Parse("demotape://type?text=roll+it+back");
        Assert.Equal("roll it back", c!.Text);
    }

    [Fact]
    public void ParseTypeRequiresText()
    {
        Assert.Null(Parse("demotape://type?cps=10"));
        Assert.Null(Parse("demotape://type?text="));
    }

    [Fact]
    public void ParseTypeClampsNegativeRate()
    {
        var c = Parse("demotape://type?text=abc&cps=-5");
        Assert.Equal("abc", c!.Text);
        Assert.Equal(0, c.Cps);
    }

    // --- typing activity ---

    [Fact]
    public void ParseTypingActivity()
    {
        var c = Parse("demotape://typing?chars=42&cps=14");
        Assert.Equal(DemoControl.CommandKind.TypingActivity, c!.Kind);
        Assert.Equal(42, c.Chars);
        Assert.Equal(14, c.Cps);
        Assert.Null(c.Caret);
    }

    [Fact]
    public void TypingActivityRateIsOptional()
    {
        var c = Parse("demotape://typing?chars=8");
        Assert.Equal(8, c!.Chars);
        Assert.Equal(0, c.Cps);
    }

    [Fact]
    public void TypingActivityRequiresPositiveCount()
    {
        Assert.Null(Parse("demotape://typing?chars=0"));
        Assert.Null(Parse("demotape://typing?cps=14"));
    }

    [Fact]
    public void TypeCarriesExpectedApp()
    {
        var c = Parse("demotape://type?text=hi&app=Chromium");
        Assert.Equal("Chromium", c!.ExpectedApp);
    }

    [Fact]
    public void ExpectedAppSurvivesPlusDecoding()
    {
        var c = Parse("demotape://type?text=hi&app=Google+Chrome");
        Assert.Equal("Google Chrome", c!.ExpectedApp);
    }

    [Fact]
    public void TypingActivityCarriesCaret()
    {
        var c = Parse("demotape://typing?chars=20&cps=14&x=420&y=680");
        Assert.Equal(new DemoControl.Point(420, 680), c!.Caret);
    }

    [Fact]
    public void CaretNeedsBothCoordinates()
    {
        var c = Parse("demotape://typing?chars=20&x=420");
        Assert.Null(c!.Caret);
    }

    // --- cursor path ---

    [Fact]
    public void ParsePathWithSeveralPoints()
    {
        var c = Parse("demotape://cursor/path?pts=120,80;200,140;260,90&ms=1400");
        Assert.Equal(DemoControl.CommandKind.CursorPath, c!.Kind);
        Assert.Equal(new[]
        {
            new DemoControl.Point(120, 80), new DemoControl.Point(200, 140), new DemoControl.Point(260, 90)
        }, c.PathPoints);
        Assert.Equal(1400, c.PathMs);
    }

    [Fact]
    public void PathDurationDefaultsWhenAbsent()
    {
        var c = Parse("demotape://cursor/path?pts=10,10;20,20");
        Assert.Equal(900, c!.PathMs);
    }

    [Fact]
    public void PathDurationHasAFloorSoItCannotFlash()
    {
        var c = Parse("demotape://cursor/path?pts=10,10;20,20&ms=5");
        Assert.Equal(120, c!.PathMs);
    }

    [Fact]
    public void PathNeedsAtLeastTwoPoints()
    {
        Assert.Null(Parse("demotape://cursor/path?pts=10,10"));
        Assert.Null(Parse("demotape://cursor/path?pts="));
        Assert.Null(Parse("demotape://cursor/path?ms=900"));
    }

    [Fact]
    public void PathSkipsMalformedPairsRatherThanFailing()
    {
        var c = Parse("demotape://cursor/path?pts=10,10;bad;30,30&ms=500");
        Assert.Equal(new[] { new DemoControl.Point(10, 10), new DemoControl.Point(30, 30) }, c!.PathPoints);
        Assert.Equal(500, c.PathMs);
    }
}
